File: android/web/src/main/java/com/prince/web/WebIntegration.java
package com.prince.web;

import android.net.Uri;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.widget.FrameLayout;

import androidx.annotation.Nullable;

import com.prince.kbd.core.CommandSpec;
import com.prince.kbd.core.IntegrationContext;
import com.prince.kbd.core.IntegrationSession;
import com.prince.kbd.core.KeyboardIntegration;

import java.util.Arrays;
import java.util.List;

/**
 * Web integration. Contributes {@code /web <url-or-query>}: mounts a {@link WebViewPanel}
 * above the keys with the resolved URL loaded.
 */
public final class WebIntegration implements KeyboardIntegration {

    private static final int PANEL_HEIGHT_DP = 440;

    @Override public String id() { return "web"; }

    @Override
    @Nullable
    public IntegrationSession activate(EditorInfo info, IntegrationContext ctx) {
        return null;
    }

    @Override
    public List<CommandSpec> commands() {
        return Arrays.asList(
                new CommandSpec("web", "Web", "🌐", true, this::handleWeb)
        );
    }

    private void handleWeb(String prompt, IntegrationContext ctx) {
        if (prompt == null || prompt.trim().isEmpty()) {
            ctx.showBanner("Try /web wikipedia.org", 1500L);
            return;
        }
        String url = resolveUrl(prompt.trim());
        WebViewPanel panel = new WebViewPanel(ctx.appContext());
        int hPx = (int) (PANEL_HEIGHT_DP * ctx.appContext().getResources().getDisplayMetrics().density);
        panel.setLayoutParams(new FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT, hPx));
        panel.setListener(ctx::hidePanel);
        ctx.showPanel(panel);
        panel.load(url);
    }

    /** URL-shaped input (has a dot, no spaces) loads directly; otherwise Google search. */
    private static String resolveUrl(String input) {
        boolean looksLikeUrl = input.contains(".") && !input.contains(" ");
        if (looksLikeUrl) {
            return input.startsWith("http://") || input.startsWith("https://")
                    ? input : "https://" + input;
        }
        return "https://www.google.com/search?q=" + Uri.encode(input);
    }
}

File: ios/TurtleKeyboardExtension/Integration/Web/WebIntegration.swift
import Foundation
#if os(iOS)
import UIKit

// MARK: - WebIntegration
//
// Mirrors android/web/WebIntegration. Ships `/web <url-or-query>` — the
// command bar collects the prompt, this handler mounts a full-panel
// WKWebView above the keys via `IntegrationContext.showPanel(...)`.
//
// No persistence, no AI, no chip — the simplest module shape.

final class WebIntegration: KeyboardIntegration {

    let id = "web"

    private static let emptyBannerMs = 1_500

    func commands() -> [CommandSpec] {
        [
            CommandSpec(
                name: "web", label: "Web", emoji: "🌐", needsPrompt: true,
                handler: { prompt, ctx in Self.handle(prompt: prompt, ctx: ctx) }
            ),
        ]
    }

    static func handle(prompt: String, ctx: IntegrationContext) {
        let trimmed = prompt.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            ctx.showBanner("Try /web wikipedia.org", autoHideMs: emptyBannerMs)
            return
        }
        guard let url = resolveURL(trimmed) else {
            ctx.showBanner("Couldn't parse that URL", autoHideMs: emptyBannerMs)
            return
        }

        let panel = WebPanelView()
        panel.setHandlers(
            onClose: { [weak ctx] in ctx?.hidePanel() },
            onOpenExternal: { [weak ctx] externalURL in
                ctx?.openExternalURL(externalURL)
            }
        )
        ctx.showPanel(panel)
        panel.load(url: url)
    }

    /// URL-ish (contains a dot, no spaces) → load directly; otherwise Google search.
    /// Matches android/web/WebIntegration.resolveUrl.
    private static func resolveURL(_ input: String) -> URL? {
        let looksLikeURL = input.contains(".") && !input.contains(" ")
        if looksLikeURL {
            let withScheme = (input.hasPrefix("http://") || input.hasPrefix("https://"))
                ? input
                : "https://\(input)"
            return URL(string: withScheme)
        }
        let encoded = input.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? input
        return URL(string: "https://www.google.com/search?q=\(encoded)")
    }

}
#endif
